File: source/python/232. Implement Queue using Stacks.py
class MyQueue:

    def __init__(self):
        self.input: list[int] = []
        self.output: list[int] = []

    def push(self, x: int) -> None:
        while self.output:
            self.input.append(self.output.pop())
        self.input.append(x)

    def pop(self) -> int:
        while self.input:
            self.output.append(self.input.pop())
        return self.output.pop()

    def peek(self) -> int:
        while self.input:
            self.output.append(self.input.pop())
        return self.output[-1]

    def empty(self) -> bool:
        return not self.input and not self.output


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()
if __name__ == "__main__":
    queue = MyQueue()

    assert queue.empty() == True

    queue.push(1)
    queue.push(2)

    assert queue.peek() == 1
    assert queue.pop() == 1
    assert queue.empty() == False

    print("All tests passed!")

File: source/cpp/232. Implement Queue using Stacks.cpp
#include <iostream>
#include <stack>


class MyQueue {
private:
    std::stack<int> input;
    std::stack<int> output;
public:
    MyQueue() {
        std::stack<int> input = std::stack<int>();
        std::stack<int> output = std::stack<int>();
    }
    
    void push(int x) {
        while (!output.empty()) {
            input.push(output.top());
            output.pop();
        }

        input.push(x);
    }
    
    int pop() {
        while (!input.empty()) {
            output.push(input.top());
            input.pop();
        }

        int result = output.top();
        output.pop();
        return result;
    }
    
    int peek() {
        while (!input.empty()) {
            output.push(input.top());
            input.pop();
        }

        return output.top();
    }
    
    bool empty() {
        return input.empty() && output.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */